### PR TITLE
chore: update changelog for v0.4

### PR DESCRIPTION
**Pull request by vlopes11**

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -1,5 +1,32 @@
...
+- Improved the execution to print the VM cycles upon termination
+
+#### Assembly
+- Added instructions: clk, ext2add, ext2sub, ext2mul, ext2div, ext2neg, ext2inv, fri_ext2fold4, hash, u32checked_popcnt, u32unchecked_popcnt
+- Added constants functionality
+- Allow empty blocks for program parse
+- Refactored `mtree_cwm` for compatibility with `mtree_set`
+
+#### Processor
+- Introduced `ProgramInfo` to bind the Kernel to a program hash
```
Comment (bobbinth) [inline CHANGELOG.md:21]: I'd change the title to `Processor, Prover, Verifier` and the content to:

- [BREAKING] Refactored `execute()`, `prove()`, `verify()` functions to take `StackInputs` as one of the parameters.
- [BREAKING] Refactored `prove()` function to return `ExecutionProof` (which is a wrapper for `StarkProof`).
- [BREAKING] Refactored `verify()` function to take `ProgramInfo`, `StackInputs`, and `ExecutionProof` as parameters and return a `u32` indicating security level of the verified proof.

**Review by bobbinth — CHANGES_REQUESTED** [review]
Thank you! Looks good! I've added a few comments with language improvements inline.

File: CHANGELOG.md [hunk 2]
```diff
@@ -1,5 +1,32 @@
 # Changelog
 
+## 0.4.0 (TBD)
+
+#### Advice provider
+- Decorator for multiplicative inverse and iNTT over quadratic extension field for the advice tape
+- Advice provider as a trait that can be injected to program execution
+- Increased flexibility to input data to the advice provider via the created `AdviceSource`
```
[the remark this answers is not in the document]
Comment (vlopes11) [reply CHANGELOG.md]: Done!

File: CHANGELOG.md [hunk 3]
```diff
@@ -1,5 +1,32 @@
...
+
+#### CLI
+- Added debug subcommand for Assembly programs with forward and backward stepping
+- Improved the execution to print the VM cycles upon termination
```
[the remark this answers is not in the document]
Comment (vlopes11) [reply CHANGELOG.md]: Done!

File: CHANGELOG.md [hunk 4]
```diff
@@ -1,5 +1,32 @@
...
+
+#### Assembly
+- Added instructions: clk, ext2add, ext2sub, ext2mul, ext2div, ext2neg, ext2inv, fri_ext2fold4, hash, u32checked_popcnt, u32unchecked_popcnt
+- Added constants functionality
+- Allow empty blocks for program parse
+- Refactored `mtree_cwm` for compatibility with `mtree_set`
```
[the remark this answers is not in the document]
Comment (vlopes11) [reply CHANGELOG.md]: Done!

File: CHANGELOG.md [hunk 5]
```diff
@@ -1,5 +1,32 @@
...
+
+#### Processor
+- Introduced `ProgramInfo` to bind the Kernel to a program hash
+
+#### Stdlib
+- Added `mem::memcopy` and `math::ext2::inv` procedures
+
+#### VM
+- Migrate hash function to Rescue Prime Optimized
+- Upgrade Winterfell backend to 0.5.1
```
[the remark this answers is not in the document]
Comment (vlopes11) [reply CHANGELOG.md]: Done!

File: CHANGELOG.md [hunk 6]
```diff
@@ -1,5 +1,32 @@
...
+- Added instructions: clk, ext2add, ext2sub, ext2mul, ext2div, ext2neg, ext2inv, fri_ext2fold4, hash, u32checked_popcnt, u32unchecked_popcnt
+- Added constants functionality
+- Allow empty blocks for program parse
+- Refactored `mtree_cwm` for compatibility with `mtree_set`
+
+#### Processor
+- Introduced `ProgramInfo` to bind the Kernel to a program hash
+
+#### Stdlib
+- Added `mem::memcopy` and `math::ext2::inv` procedures
```
[the remark this answers is not in the document]
Comment (vlopes11) [reply CHANGELOG.md]: Done!

File: CHANGELOG.md (hunk 1 above)
Comment (vlopes11) [reply CHANGELOG.md]: Done!

**Review by bobbinth — APPROVED** [review]
Looks good! Thank you!

**Pull request merged**